File: java/base/src/main/vsdk/toolkit/gui/tangibleInterfaces/TangibleInterfaceEvent.java
package vsdk.toolkit.gui.tangibleInterfaces;

import vsdk.toolkit.common.linealAlgebra.Quaterniond;
import vsdk.toolkit.common.linealAlgebra.Vector3Dd;
import vsdk.toolkit.gui.PresentationElement;

/**
This class represents a single 6-DoF pose update reported by a tangible
interface marker tracking service (see `TangibleInterfaceNetworkClient`),
identifying a marker group, its position and its rotation.
*/
public class TangibleInterfaceEvent extends PresentationElement {
    private String id;
    private Vector3Dd position;
    private Quaterniond rotation;

    /**
    Builds a tangible interface event.

    @param id identifier (label) of the marker group reporting the pose
    @param position position of the marker group, in meters
    @param rotation orientation of the marker group
    */
    public TangibleInterfaceEvent(String id, Vector3Dd position, Quaterniond rotation) {
        this.id = id;
        this.position = position;
        this.rotation = rotation;
    }

    /**
    @return identifier (label) of the marker group reporting the pose
    */
    public String getId() {
        return id;
    }

    /**
    @return position of the marker group, in meters
    */
    public Vector3Dd getPosition() {
        return position;
    }

    /**
    @return orientation of the marker group
    */
    public Quaterniond getRotation() {
        return rotation;
    }

    public String toString() {
        return "TangibleInterfaceEvent{id=" + id + ", position=" + position + ", rotation=" + rotation + "}";
    }
}

File: typescript/base/src/main/vsdk/toolkit/gui/tangibleInterfaces/TangibleInterfaceEvent.ts
import { Quaterniond } from "../../common/linealAlgebra/Quaterniond.js";
import { Vector3Dd } from "../../common/linealAlgebra/Vector3Dd.js";
import { PresentationElement } from "../PresentationElement.js";

/** A 6-DoF pose update emitted by a tangible-marker tracking service. */
export class TangibleInterfaceEvent extends PresentationElement {
    public constructor(
        private readonly id: string,
        private readonly position: Vector3Dd,
        private readonly rotation: Quaterniond,
    ) {
        super();
    }
    public getId(): string {
        return this.id;
    }
    public getPosition(): Vector3Dd {
        return this.position;
    }
    public getRotation(): Quaterniond {
        return this.rotation;
    }
    public override toString(): string {
        return `TangibleInterfaceEvent{id=${this.id}, position=${this.position}, rotation=${this.rotation}}`;
    }
}
